dev/TabView/TabView.cpp

```diff
@@ -0,0 +1,249 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#include "pch.h"
+#include "common.h"
+#include "TabView.h"
+#include "TabViewAutomationPeer.h"
+#include "DoubleUtil.h"
+#include "RuntimeProfiler.h"
+#include "ResourceAccessor.h"
+#include "SharedHelpers.h"
+
+static constexpr double c_tabMinimumWidth = 48.0;
+static constexpr double c_tabMaximumWidth = 200.0;
+
+static constexpr wstring_view c_tabViewItemMinWidthName{ L"TabViewItemMinWidth"sv };
+static constexpr wstring_view c_tabViewItemMaxWidthName{ L"TabViewItemMaxWidth"sv };
+
+// TODO: what is the right number and should this be customizable?
+static constexpr double c_scrollAmount = 50.0;
+
+TabView::TabView()
+{
+    __RP_Marker_ClassById(RuntimeProfiler::ProfId_TabView);
+
+    SetDefaultStyleKey(this);
+
+    Loaded({ this, &TabView::OnLoaded });
+    SelectionChanged({ this, &TabView::OnSelectionChanged });
+    SizeChanged({ this, &TabView::OnSizeChanged });
+}
+
+void TabView::OnApplyTemplate()
+{
+    winrt::IControlProtected controlProtected{ *this };
+
+    m_tabContentPresenter.set(GetTemplateChildT<winrt::ContentPresenter>(L"TabContentPresenter", controlProtected));
+
+    m_scrollViewer.set([this, controlProtected]() {
+        auto scrollViewer = GetTemplateChildT<winrt::FxScrollViewer>(L"ScrollViewer", controlProtected);
+        if (scrollViewer)
+        {
+            m_scrollViewerLoadedRevoker = scrollViewer.Loaded(winrt::auto_revoke, { this, &TabView::OnScrollViewerLoaded });
+        }
+        return scrollViewer;
+    }());
+
+    m_scrollViewer.set(GetTemplateChildT<winrt::FxScrollViewer>(L"ScrollViewer", controlProtected));
+    if (auto scrollViewer = m_scrollViewer.get())
+    {
+        m_scrollViewerLoadedRevoker = scrollViewer.Loaded(winrt::auto_revoke, { this, &TabView::OnScrollViewerLoaded });
+    }
+}
```

Comment:
I think this code is repeated and can be removed?

Also I like this new pattern you have above, this feels like the pattern I've been searching for where OnApplyTemplate can look for parts that might not be there but still unconditionally set the field (so if it's not found the field ends up as nullptr).

Reply:
although... the revokers would still not be cleared, so it still needs a little something else.

---
In reply to: [289588743](https://github.com/microsoft/microsoft-ui-xaml/pull/629#discussion_r289588743) [](ancestors = 289588743)

Reply:
oops! I didn't remove the old code.